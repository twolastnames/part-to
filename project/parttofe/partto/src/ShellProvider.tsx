import React, { MutableRefObject, useRef, useState } from "react";

import { MantineProvider, createTheme } from "@mantine/core";
import { PropsWithChildren } from "react";
import "@mantine/core/styles.layer.css";
import "./App.scss";

const themes = {
  default: createTheme({
    primaryColor: "red",
  }),
  dark: createTheme({
    primaryColor: "red",
  }),
};

let themeChangeRef: MutableRefObject<(theme: Theme) => void>;

export enum Theme {
  Default = "default",
  Dark = "dark",
}

export function changeTheme(theme: Theme) {
  const themeNode = document.getElementById("theme");
  if (!themeNode || !themeNode.className) {
    console.error("developers should have created an theme ID in the project");
    return;
  }
  themeChangeRef?.current && themeChangeRef.current(theme);
  themeNode.className = `${theme}Theme`;
}

export const ShellProvider = ({ children }: PropsWithChildren) => {
  const [theme, setTheme] = useState<Theme>(Theme.Default);
  themeChangeRef = useRef(setTheme);
  return (
    <div id="theme" className="defaultTheme">
      <MantineProvider theme={themes[theme]}>{children}</MantineProvider>
    </div>
  );
};
