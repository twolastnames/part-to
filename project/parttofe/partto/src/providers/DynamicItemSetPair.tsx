import React, {
  createContext,
  useState,
  Context,
  MutableRefObject,
  createRef,
  useContext,
  SetStateAction,
} from "react";
import { PropsWithChildren } from "react";
import { Duration, getDuration } from "../shared/duration";
import { debounce, DebouncedFunc } from "lodash";

export type DynamicItemSetPairType = {
  selected: number;
  setSelected: (selected: number) => void;
  getTotal: () => number;
  goBack: () => void;
  goForward: () => void;
  showDuration: Duration;
  togglePause: () => void;
  paused: boolean;
  nextShowDuration: () => void;
};

export const UndefinedDynamicItemSetPair = {
  selected: 0,
  setSelected: (_: number) => undefined,
  goBack: () => undefined,
  goForward: () => undefined,
  getTotal: () => 0,
  showDuration: getDuration(0),
  togglePause: () => undefined,
  paused: true,
  nextShowDuration: () => undefined,
};

export function useSingleOfPair({ context }: ContextDescription) {
  return useContext(context) || UndefinedDynamicItemSetPair;
}

type ContextType = Context<undefined | DynamicItemSetPairType>;

export type ContextDescription = {
  context: ContextType;
  setCount: (value: number) => void;
  setSettingKey: (settingKey: string) => void;
  setDurations: (value: Array<Duration>) => void;
};

const defaultDurations = [1000, 2000, 4000, 8000];

const getDurations = (key: string) =>
  JSON.parse(
    localStorage.getItem(`${key}Durations`) || JSON.stringify(defaultDurations),
  ).map(getDuration);

const setDurations = (key: string, value: Array<Duration>) => {
  localStorage.setItem(
    `${key}Durations`,
    JSON.stringify(value.map((duration) => duration.toMilliseconds())),
  );
};

const leftCount: MutableRefObject<((value: number) => void) | null> =
  createRef();
const leftSettingKey: MutableRefObject<string | null> = createRef();
const getLeftSettingKey = () => leftSettingKey.current || "left";
function setLeftCount(value: number) {
  leftCount.current && leftCount.current(value);
}
const LeftDynamicItemSetPairContext: ContextType = createContext<
  undefined | DynamicItemSetPairType
>(undefined);

export const LeftContext: ContextDescription = {
  context: LeftDynamicItemSetPairContext,
  setCount: setLeftCount,
  setSettingKey: (value: string) => {
    leftSettingKey.current = value;
  },
  setDurations: (value: Array<Duration>) => {
    setDurations(getLeftSettingKey(), value);
  },
};

const rightCount: MutableRefObject<((value: number) => void) | null> =
  createRef();
const rightSettingKey: MutableRefObject<string | null> = createRef();
const getRightSettingKey = () => leftSettingKey.current || "left";

function setRightCount(value: number) {
  rightCount.current && rightCount.current(value);
}
const RightDynamicItemSetPairContext: ContextType = createContext<
  undefined | DynamicItemSetPairType
>(undefined);

export const RightContext: ContextDescription = {
  context: RightDynamicItemSetPairContext,
  setCount: setRightCount,
  setSettingKey: (value: string) => {
    rightSettingKey.current = value;
  },
  setDurations: (value: Array<Duration>) => {
    setDurations(getRightSettingKey(), value);
  },
};

type GoADirection = (
  totalPages: number,
  setSelected: (mutator: (arg: ProviderState) => ProviderState) => void,
) => void;

const goBack: GoADirection = (length, setState) => {
  setState((last: ProviderState) => ({
    ...last,
    selected: last.selected <= 0 ? length - 1 : last.selected - 1,
  }));
};

const goForward: GoADirection = (length, setState) => {
  setState((last: ProviderState) => ({
    ...last,
    selected: last.selected >= length - 1 ? 0 : last.selected + 1,
  }));
};

type ProviderState = {
  selected: number;
  showDuration: number;
  paused: boolean;
  count: number;
};

function useProvider({
  setCount,
  getSettingKey,
  togglePauseClickSlower,
}: {
  getSettingKey: () => string;
  setCount: MutableRefObject<((value: number) => void) | null>;
  togglePauseClickSlower: DebouncedFunc<
    (setState: (arg: SetStateAction<ProviderState>) => void) => void
  >;
}) {
  const durations = getDurations(getSettingKey());
  const [{ count, selected, showDuration, paused }, setState] =
    useState<ProviderState>({
      selected: 0,
      showDuration: Math.floor(durations.length / 2),
      paused: false,
      count: 0,
    });

  setCount.current = (value: number) => {
    if (value === count) {
      return;
    }
    setState((previous) => ({ ...previous, count: value }));
  };

  if (selected !== 0 && (selected < 0 || selected >= count)) {
    setState((current) => ({ ...current, selected: 0 }));
  }

  const duration = durations[showDuration];
  return {
    selected: selected,
    setSelected: (selected: number) => {
      setState((current) => ({ ...current, selected }));
    },
    goBack: () => {
      goBack(count, setState);
    },
    goForward: () => {
      goForward(count, setState);
    },
    showDuration: duration,
    paused: paused,
    getTotal: () => count,
    togglePause: () => {
      togglePauseClickSlower(setState);
    },
    nextShowDuration: () => {
      const next = showDuration + 1;
      setState((previous) => ({
        ...previous,
        showDuration: next >= durations.length ? 0 : next,
      }));
    },
  };
}

const leftTogglePauseClickSlower = debounce(
  (setState: (arg: SetStateAction<ProviderState>) => void) => {
    setState((previous) => ({ ...previous, paused: !previous.paused }));
  },
  100,
);

export function LeftDynamicItemSetPairProvider({
  children,
}: PropsWithChildren) {
  return (
    <LeftDynamicItemSetPairContext.Provider
      value={useProvider({
        setCount: leftCount,
        getSettingKey: getLeftSettingKey,
        togglePauseClickSlower: leftTogglePauseClickSlower,
      })}
    >
      {children}
    </LeftDynamicItemSetPairContext.Provider>
  );
}

const rightTogglePauseClickSlower = debounce(
  (setState: (arg: SetStateAction<ProviderState>) => void) => {
    setState((previous) => ({ ...previous, paused: !previous.paused }));
  },
  100,
);

export function RightDynamicItemSetPairProvider({
  children,
}: PropsWithChildren) {
  return (
    <RightDynamicItemSetPairContext.Provider
      value={useProvider({
        setCount: rightCount,
        getSettingKey: getRightSettingKey,
        togglePauseClickSlower: rightTogglePauseClickSlower,
      })}
    >
      {children}
    </RightDynamicItemSetPairContext.Provider>
  );
}
