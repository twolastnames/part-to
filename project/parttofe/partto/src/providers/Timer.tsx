import React, {
  createContext,
  createRef,
  MutableRefObject,
  useContext,
  useState,
} from "react";
import { PropsWithChildren } from "react";
import { DateTime } from "../shared/dateTime";
import { Duration } from "../shared/duration";

export type TimerType = {
  timers: TimerSet;
  add: (id: string, start: DateTime, duration: Duration) => void;
  clear: () => void;
};

type TimerDescription = {
  start: DateTime;
  duration: Duration;
};

export interface TimerSet {
  [key: string]: TimerDescription;
}

export const timerClearer: MutableRefObject<undefined | null | (() => void)> =
  createRef<undefined | (() => void)>();

export const clearTimers = () => {
  timerClearer.current && timerClearer.current();
};

export const TimerContext = createContext<undefined | TimerType>(undefined);

export function TimerProvider({ children }: PropsWithChildren) {
  const [timers, setTimers] = useState<TimerSet>({});
  timerClearer.current = () => {
    setTimers({});
  };
  return (
    <TimerContext.Provider
      value={{
        timers,
        add: (id, start, duration) => {
          if (
            Object.keys(timers).includes(id) &&
            timers[id].duration === duration &&
            timers[id].start === start
          ) {
            return;
          }
          setTimers((previous) => ({ ...previous, [id]: { start, duration } }));
        },
        clear: () => {
          setTimers({});
        },
      }}
    >
      {children}
    </TimerContext.Provider>
  );
}

export function useTimerProvider() {
  return (
    useContext(TimerContext) || {
      timers: {},
      add: () => undefined,
    }
  );
}
