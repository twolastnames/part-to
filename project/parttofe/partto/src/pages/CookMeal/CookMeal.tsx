import React from "react";

import { Layout } from "../../components/Layout/Layout";
import { ManageTasksIdFromer } from "../../components/ManageTasks/ManageTasks";
import { useParams } from "react-router-dom";
import { LeftContext, RightContext } from "../../providers/DynamicItemSetPair";

type Entity = {
  name: string;
  side: string;
};

const getEmptyText = (target: Entity, other: Entity) =>
  [
    `No current ${target.name}. If any ${target.name} are left,`,
    `at least some the ${other.name}`,
    `on the ${other.side} pane need to be finished first.`,
  ].join(" ");

export function CookMeal() {
  const { runState } = useParams() as { runState: string };
  return (
    <Layout
      pair={[
        <ManageTasksIdFromer
          runState={runState}
          typeKey="duties"
          context={LeftContext}
          emptyText={getEmptyText(
            { name: "duties", side: "first" },
            { name: "tasks", side: "second" },
          )}
        />,
        <ManageTasksIdFromer
          runState={runState}
          context={RightContext}
          typeKey="tasks"
          emptyText={getEmptyText(
            { name: "tasks", side: "second" },
            { name: "duties", side: "first" },
          )}
        />,
      ]}
    />
  );
}
