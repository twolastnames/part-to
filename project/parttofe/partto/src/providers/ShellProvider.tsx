import React, { MutableRefObject, useRef, useState } from "react";

import { MantineProvider, createTheme } from "@mantine/core";
import { PropsWithChildren } from "react";
import "@mantine/core/styles.layer.css";
import "../App.scss";
import {
  LeftDynamicItemSetPairProvider,
  RightDynamicItemSetPairProvider,
} from "./DynamicItemSetPair";
import { TimerProvider } from "./Timer";

export enum Theme {
  Default = "default",
  Dark = "dark",
}

const themeStorageKey = "theme";

export function getTheme(): Theme {
  return (localStorage.getItem(themeStorageKey) || Theme.Default) as Theme;
}

const themes = {
  default: createTheme({
    primaryColor: "red",
  }),
  dark: createTheme({
    primaryColor: "red",
  }),
};

let themeChangeRef: MutableRefObject<(theme: Theme) => void>;

function setThemeClass() {
  const themeNode = document.getElementById("theme");
  if (!themeNode || !themeNode.className) {
    console.error("developers should have created an theme ID in the project");
    return;
  }
  themeNode.className = `${getTheme()}Theme`;
}

export function changeTheme(theme: Theme) {
  localStorage.setItem(themeStorageKey, theme.toString());
  themeChangeRef?.current && themeChangeRef.current(theme);
}

export const ShellProvider = ({ children }: PropsWithChildren) => {
  const [theme, setTheme] = useState<Theme>(getTheme());
  themeChangeRef = useRef(setTheme);
  setTimeout(() => {
    setThemeClass();
  }, 0);
  return (
    <div id="theme" className="defaultTheme">
      <LeftDynamicItemSetPairProvider>
        <RightDynamicItemSetPairProvider>
          <TimerProvider>
            <MantineProvider theme={themes[theme]}>{children}</MantineProvider>
          </TimerProvider>
        </RightDynamicItemSetPairProvider>
      </LeftDynamicItemSetPairProvider>
    </div>
  );
};
