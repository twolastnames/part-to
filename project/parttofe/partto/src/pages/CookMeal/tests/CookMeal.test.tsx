import React from "react";
import { expect, test } from "@jest/globals";
import { render, screen } from "@testing-library/react";
import { ShellProvider } from "../../../providers/ShellProvider";
import { CookMeal } from "../CookMeal";

test("snapshot", () => {
  render(
    <ShellProvider>
      <CookMeal />
    </ShellProvider>,
  );
  const page = screen.getByTestId("Layout");
  expect(page).toMatchSnapshot();
});
