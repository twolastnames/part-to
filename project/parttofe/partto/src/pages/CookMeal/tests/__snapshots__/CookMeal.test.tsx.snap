// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  class="layout"
  data-testid="Layout"
>
  <div
    class="itemSet"
  >
    <div
      class="error"
      data-testid="Spinner"
    >
      Loading...
    </div>
  </div>
  <div
    class="itemSet"
  >
    <div
      class="error"
      data-testid="Spinner"
    >
      Loading...
    </div>
  </div>
</div>
`;
